diff -ru gdb-5.2.orig/gdb/Makefile.in gdb-5.2/gdb/Makefile.in
--- gdb-5.2.orig/gdb/Makefile.in	Mon Feb 25 19:15:52 2002
+++ gdb-5.2/gdb/Makefile.in	Thu May 30 22:33:31 2002
@@ -548,7 +548,8 @@
 	tui/tui-file.h tui/tui-file.c tui/tui-out.c tui/tui-hooks.c \
 	ui-file.h ui-file.c \
 	frame.c doublest.c \
-	gnu-v2-abi.c gnu-v3-abi.c hpacc-abi.c cp-abi.c
+	gnu-v2-abi.c gnu-v3-abi.c hpacc-abi.c cp-abi.c \
+	csharp-lang.c csharp-mono-lang.c
 
 LINTFILES = $(SFILES) $(YYFILES) $(CONFIG_SRCS) init.c
 
@@ -670,7 +671,8 @@
 	vx-share/dbgRpcLib.h vx-share/ptrace.h vx-share/vxTypes.h \
 	vx-share/vxWorks.h vx-share/wait.h vx-share/xdr_ld.h \
 	vx-share/xdr_ptrace.h vx-share/xdr_rdb.h gdbthread.h \
-	dcache.h remote-utils.h top.h somsolib.h
+	dcache.h remote-utils.h top.h somsolib.h \
+	csharp-lang.h
 
 # Header files that already have srcdir in them, or which are in objdir.
 
@@ -722,7 +724,8 @@
 	nlmread.o serial.o mdebugread.o os9kread.o top.o utils.o \
 	ui-file.o \
 	frame.o doublest.o \
-	gnu-v2-abi.o gnu-v3-abi.o hpacc-abi.o cp-abi.o
+	gnu-v2-abi.o gnu-v3-abi.o hpacc-abi.o cp-abi.o \
+	csharp-lang.o csharp-mono-lang.o
 
 OBS = $(COMMON_OBS) $(ANNOTATE_OBS)
 
diff -ru gdb-5.2.orig/gdb/defs.h gdb-5.2/gdb/defs.h
--- gdb-5.2.orig/gdb/defs.h	Mon Mar 25 17:50:20 2002
+++ gdb-5.2/gdb/defs.h	Thu May 30 22:32:57 2002
@@ -209,7 +209,8 @@
     language_m2,		/* Modula-2 */
     language_asm,		/* Assembly language */
     language_scm,    		/* Scheme / Guile */
-    language_pascal		/* Pascal */
+    language_pascal,		/* Pascal */
+    language_csharp_mono	/* C# using Mono */
   };
 
 enum precision_type
diff -ru gdb-5.2.orig/gdb/dwarf2read.c gdb-5.2/gdb/dwarf2read.c
--- gdb-5.2.orig/gdb/dwarf2read.c	Thu Feb 28 12:21:16 2002
+++ gdb-5.2/gdb/dwarf2read.c	Thu May 30 22:32:57 2002
@@ -3822,6 +3822,9 @@
     case DW_LANG_Java:
       cu_language = language_java;
       break;
+    case DW_LANG_CSharp_Mono:
+      cu_language = language_csharp_mono;
+      break;
     case DW_LANG_Ada83:
     case DW_LANG_Cobol74:
     case DW_LANG_Cobol85:
diff -ru gdb-5.2.orig/gdb/language.c gdb-5.2/gdb/language.c
--- gdb-5.2.orig/gdb/language.c	Wed Feb 13 19:49:30 2002
+++ gdb-5.2/gdb/language.c	Thu May 30 22:32:57 2002
@@ -865,6 +865,7 @@
     case language_chill:
     case language_m2:
     case language_pascal:
+    case language_csharp_mono:
       return TYPE_CODE (type) != TYPE_CODE_CHAR ? 0 : 1;
 
     case language_c:
diff -ru gdb-5.2.orig/include/elf/dwarf2.h gdb-5.2/include/elf/dwarf2.h
--- gdb-5.2.orig/include/elf/dwarf2.h	Tue Jan 29 00:26:53 2002
+++ gdb-5.2/include/elf/dwarf2.h	Thu May 30 22:32:57 2002
@@ -675,7 +675,8 @@
     DW_LANG_Ada95 = 0x000d,
     DW_LANG_Fortran95 = 0x000e,
     /* MIPS.  */
-    DW_LANG_Mips_Assembler = 0x8001
+    DW_LANG_Mips_Assembler = 0x8001,
+    DW_LANG_CSharp_Mono = 0x9001
   };
 
 
