diff --git a/src/absil/ilwrite.fs b/src/absil/ilwrite.fs
index 8ea9dec..74b3d9d 100755
--- a/src/absil/ilwrite.fs
+++ b/src/absil/ilwrite.fs
@@ -2594,7 +2594,8 @@ let GenMethodDefAsRow cenv env midx (md: ILMethodDef) =
                 SequencePoints=seqpoints }
           cenv.AddCode code
           addr
-      | MethodBody.Abstract ->
+      | MethodBody.Abstract
+      | MethodBody.PInvoke _ ->
           // Now record the PDB record for this method - we write this out later. 
           if cenv.generatePdb then 
             cenv.pdbinfo.Add  
@@ -2607,8 +2608,7 @@ let GenMethodDefAsRow cenv env midx (md: ILMethodDef) =
                 SequencePoints = [| |] }
           0x0000
       | MethodBody.Native -> 
-          failwith "cannot write body of native method - Abstract IL cannot roundtrip mixed native/managed binaries"
-      | _  -> 0x0000)
+          failwith "cannot write body of native method - Abstract IL cannot roundtrip mixed native/managed binaries")
 
     UnsharedRow 
        [| ULong  codeAddr  
