diff --git a/mono/config.make b/mono/config.make
index 00bd4a091..d7090a877 100644
--- a/mono/config.make
+++ b/mono/config.make
@@ -4,7 +4,12 @@ DEFAULT: all
 
 monocmd = $(shell which mono)
 monocmddir = $(dir $(monocmd))
+ifeq (x-$(PREFIX)-,x--)
 prefix = $(shell (cd $(monocmddir)/..; pwd))
+else
+prefix = $(PREFIX)
+endif
+
 thisdir = $(dir $(realpath $(lastword $(MAKEFILE_LIST))))
 topdir = $(thisdir)../
 builddir = $(topdir)
@@ -147,7 +152,7 @@ install-sdk-lib:
 	@mkdir -p $(DESTDIR)$(monodir)/fsharp
 	@if test "x$(DELAY_SIGN)" = "x1"; then \
 	    echo "Signing $(outdir)$(ASSEMBLY) with Mono key"; \
-	    $(monobindir)sn -q -R $(outdir)$(ASSEMBLY) $(topdir)mono/mono.snk; \
+	    $(monocmddir)sn -q -R $(outdir)$(ASSEMBLY) $(topdir)mono/mono.snk; \
 	fi
 	@if test x-$(NAME) = x-FSharp.Compiler.Private; then \
 	    echo "Installing extra dependency System.Collections.Immutable.dll to $(DESTDIR)$(monodir)/fsharp/"; \
